### PR TITLE
Added support for build arguments.

### PR DESCRIPTION
**Pull request by leonty**
This implements the feature #837

### CODE REVIEW

**Comment by leonty** [comment]
Done!

**Comment by leonty** [comment]
@shin what's the due date for the 1.6.0 milestone?

**Comment by dnephin** [comment]
Thanks for the PR!

A test for this case would be great.

I don't think we have a set release date for the 1.6.0 milestone.

**Comment by leonty** [comment]
@dnephin voilà

File: tests/integration/build_test.py [hunk 1]
```diff
@@ -96,3 +96,16 @@ def test_build_with_dockerignore(self):
...
+    @requires_api_version('1.9')
+    def test_build_with_buildargs(self):
+        script = io.BytesIO('\n'.join([
+            'FROM scratch',
+            'ARG test',
+            'USER $test'
+        ]).encode('ascii'))
+
+        self.client.build(fileobj=script, tag='buildargs',
+                          buildargs={'test': 'OK'}, stream=True)
```
Comment (shin-) [inline tests/integration/build_test.py]: The created image needs to be cleaned up after the test has finished running. `self.tmp_imgs.append('buildargs')` should do the trick.

**Comment by shin-** [comment]
Thanks!

One small comment on the test. Additionally, can you squash those commits into one? Once that's done, LGTM.

**Comment by leonty** [comment]
@shin- no problem. But I used the existing tests as a reference and they don't clean anything after (all integration build tests)...

**Comment by leonty** [comment]
@shin- done. Though I'm not sure the squash was correct, have never done it before

**Comment by leonty** [comment]
Hi, is this accepeted? I'm about to push another PR for Ansible to use this functionality and just waiting to know for sure that this API will be in 1.6.

**Comment by shin-** [comment]
Thanks! Looking back there's a few things still.

1. Your commit is missing the signoff at the bottom.
2. The test should require version `1.21` (the latest API version) instead of `1.9` (the engine version).
3. In `Client.build`, if `buildargs` is provided, we should check the version is greater or equal `1.21` (`utils.version_gte` is a good way to do that) and raise an `errors.InvalidVersion` if it isn't.

**Comment by leonty** [comment]
Done. But I don't know why the integration test has failed, for me locally it works:
```
>>> import docker, io
>>> c = docker.Client('127.0.0.1:4243')
>>> script = io.BytesIO('\n'.join([
            'FROM scratch',
            'ARG test',
            'USER $test'
        ]).encode('ascii'))
>>> c.build(fileobj=script, tag='buildargs', buildargs={'test': 'OK'}, stream=True)
>>> info = c.inspect_image('buildargs')
>>> info
{'Os': 'linux', 'Parent': 'bc2b3187f9a63a6bc427c2b6a80675ec4885a4262d8f8ed7b04ba14c9b848ce2', 'RepoTags': ['buildargs:latest'], 'Config': {'Hostname': '4ee80013
...
```

**Comment by shin-** [comment]
There's a race condition where your test is progressing before the build operation is finished on the server side. I've updated the PR and submitted it as #846 . Let's follow up there.

**Comment by leonty** [comment]
Ok, thanks.

**Comment by shin-** [comment]
Merged into master :+1:

**Comment by leonty** [comment]
Great! Thanks again.

**Pull request closed without merging**